controller/api/destination/endpoint_translator.go

```diff
@@ -36,10 +37,82 @@ func newEndpointTranslator(
 		"service":   service,
 	})
 
-	return &endpointTranslator{controllerNS, identityTrustDomain, enableH2Upgrade, stream, log}
+	return &endpointTranslator{controllerNS, identityTrustDomain, enableH2Upgrade, stream, log, watcher.TopologyFilter{EnableEndpointSlices: false}}
+}
+
+// TODO: what if we move this (and the topologyFilter struct) as an argument to Add
+// you add an address with a topology filter
+// might not work because how are we going to get the service topology in the deep levels
+// e.g it can work with port creation but will it work with adding a slice? don't think so
+func (et *endpointTranslator) AddTopologyFilter(filter watcher.TopologyFilter) {
+	et.log.Debugf("Adding topology filter with preference %v and node topology %v", filter.ServicePreference, filter.NodeTopology)
+	et.TopologyFilter = filter
+}
+
+func (et *endpointTranslator) filterAddresses(set watcher.AddressSet) watcher.AddressSet {
+	et.log.Debugf("Filtering through address set with preference %v", et.ServicePreference)
+	// Do filtering here
+	// filterEndpointSlices
+	// to do this, we need: service topology, node topology (addresses have their own)
+	//filtered := filterAddresses(set, topologyPref, nodeTopology)
+	//et.Add(filtered)
+	/* TODO: would be cool to keep count of how many we have?
+		for example: pref is [host, region, *] and endpoints are [{host,region}, {host}, {host, region, zone}]
+	==>
+	count: host: 3
+	       region: 2
+	       zone: 1
+	but I don't know if this is helpful at all :shrug
+	*/
+
+	filtered := make(map[watcher.ID]watcher.Address)
+	for _, locality := range et.ServicePreference {
+		// If we encounter '*' it means all endpoints regardless of topology pref
+		// makes no sense to keep going, just break
+		if locality == "*" {
+			break
+		}
+		et.log.Debugf("Locality [%s]", locality)
+		// If, for some reason, the src node does not have the currently
+		// preferred topology, means we skip this
+		//TODO: can this happen in effect? thought all nodes have those labels!
+		nodeLocality, ok := et.NodeTopology[locality]
+		if !ok {
+			continue
+		}
+
+		// Nested for-loop, outer loop has at most 4 items input should not
+		// grow out of proportion
+		for id, address := range set.Addresses {
+			et.log.Debugf("Address [%s:%s]: Topology is %v ", address.IP, address.Pod.Name, address.Topology)
+			// No need to test if it exists, Go gives us zero-values
+			addrLocality := address.Topology[locality]
+			if addrLocality == nodeLocality {
+				filtered[id] = address
+			}
+		}
+
+		// stop looping
+		if len(filtered) > 0 {
+			et.log.Debugf("Filtered %d from a total of %d", len(filtered), len(set.Addresses))
+			return watcher.AddressSet{
+				Addresses: filtered,
+				Labels:    set.Labels,
+			}
+		}
+	}
+
+	// If we reach this point it means we either have no satisfied topologies
+	// or we encountered the wildcard. Return in both cases what we previously had
```

Comment:
That's a very valid point of view. I thought similarily but I didn't see any easy way of sending a `NoEndpoints` message since the filtering happens in `Add` based on how I have done things. I think in this scenario we could either have the filtering done before calling `Add` or return an empty set from the filtering method. Hope I didn't miss anything.